cmake: APPLICATION_CONFIG_DIR support implemented

tejlmand opened the pull request:
With this commit a dedicated APPLICATION_CONFIG_DIR is added to the
Zephyr build system.

Currently, the APPLICATION_SOURCE_DIR is also the base
location of configuration files.

This is very practical for simple samples, but also has its limitations
for more complex setups.

Introducing a dedicated APPLICATION_CONFIG_DIR allows users more
customization options in Zephyr build system.
Especially in terms of custom build configuration files unknown to
Zephyr itself.

For example, instead of all configuration files being located directly
in the application source folder, a project might prefer to organize
their configuration files on a per board basis, for example:
<app>/boards/custom_board_A/prj.conf
<app>/boards/custom_board_A/app.overlay
<app>/boards/custom_board_A/custom_file.mine
<app>/boards/custom_board_B/prj.conf
<app>/boards/custom_board_B/app.overlay
<app>/boards/custom_board_B/custom_file.mine
...

instead of n-files located in the root of the sample.

If the user / sample specifies APPLICATION_CONFIG_DIR, then this folder will
always be used instead of the default configuration folder.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: APPLICATION_CONFIG_DIR support implemented

With this commit a dedicated APPLICATION_CONFIG_DIR is added to the
Zephyr build system.

Currently, the APPLICATION_SOURCE_DIR is identical also the base
location of configuration files.

This is very practical for simple samples, but also has it limitations
for more complex setups.

Introducing a dedicated APPLICATION_CONFIG_DIR allows users more
customization options in Zephyr build system.
Especially in terms of custom build configuration files unknown to
Zephyr itself.

For example, instead of all configuration files being located directly
in the application source folder, a project might prefer to organize
their configuration files on a per board basis, for example:
<app>/boards/custom_board_A/prj.conf
<app>/boards/custom_board_A/app.overlay
<app>/boards/custom_board_A/custom_file.mine
<app>/boards/custom_board_B/prj.conf
<app>/boards/custom_board_B/app.overlay
<app>/boards/custom_board_B/custom_file.mine
...

instead of n-files located in the root of the sample.

If the user / sample specifies APPLICATION_CONFIG_DIR, then this folder
will always be used instead of the default configuration folder.

As part of this extension the behaviour of
`-DCONF_FILE=<relative-path>/prj_<build>.conf` and additional Kconfig
fragments automatic sourcing has been aligned with the default behavior
of `prj.conf`.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

mbolivar-nordic commented:
A configurable APPLICATION_CONFIG_DIR feels like a first step towards the Zephyr Development 2021 discussion about having multiple places throughout the tree where we can pick up .overlay and .conf files, as pushed for by @JordanYates. So I like the direction this is going.

JordanYates commented:
I can see this being useful generally, but I don't think it's movement towards picking up common files for boards.
Common overlays for different applications still need to be copy-pasted into each application.
I know that this PR isn't intended to solve that problem, but I'm not sure how it could be extended to help either.

tejlmand commented:
> Common overlays for different applications still need to be copy-pasted into each application.

Yes, such feature is outside the scope of this PR.
And regarding such feature as sharing Kconfig fragments between applications we should be careful.

Today a user can go into a sample, like `samples/net/sockets/echo_server`, and find all needed configs that will work with that sample.

If multiple samples start sharing Kconfig fragments, such as: `overlay-802154.conf` from another folder, lets say `config/common/`, then as a user I need to consider: `Is this Kconfig fragment compatible with sampleA that i'm building ?`

It makes it harder for a user to create a custom app based on an existing sample.
Today a user may simple do:
```
$ cp sample/<name> <my_repo>/app
# Adjust and extend sample to personal needs
```
but if we start placing Kconfig fragments in multiple places, users need to:
```
$ cp sample/<name> <my_repo>/app
# Find additional Kconfig fragments from folder A, B, and C
$ cp <extra-configs-dir-A>/fragment.conf <my_repo>/app
# or adjust `prj.conf` with content of said fragment.conf.

# and finally adjust and extend sample to personal needs
```

I'm not against common locations for simple settings, but we need to be careful.
Zephyr is already rather complex, and start placing additional configs at multiple locations might make it harder for users.

So we need to be very careful in such regards, and thus it is not part of this PR.

mbolivar-nordic commented on doc/guides/build/kconfig/setting.rst:
```diff
@@ -167,6 +167,22 @@ The application configuration can come from the sources below. By default,
+
+The application's configuration directory will be defined as:
+
+1. If ``APPLICATION_CONFIG_DIR`` is specified by the user with
```
I feel like this section belongs in doc/application/index.rst instead, no?

tejlmand replied:
I agree, it does.

But I also feel that this knowledge, and the knowledge regarding the prj.conf file belongs together.

Having those split can actually make it hard to understand exactly how the configuration / overlay scheme works.

Will take a second look, and if you have any good proposals, let me know.

mbolivar-nordic replied:
Why not put it in the application guide and link to it from here?

mbolivar-nordic commented:
>  I know that this PR isn't intended to solve that problem, but I'm not sure how it could be extended to help either.

My thinking here is that APPLICATION_CONFIG_DIR could be a list, and we could look in conventional places across the workspace for populating it, depending on some criteria.

mbolivar-nordic commented:
@tejlmand I've proposed this for next week's dev-review meeting as I think it's worth getting more attention on it, now that main is open again.

marinjurjevic commented:
Can I just chime in by saying whatever you decide on have it clearly documented somewhere. I like the way Nordic is doing with [multi-image build scenarios](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/ug_multi_image.html#image-specific-variables).

That being said, I would like documentation clearing up any ambiguity regarding file structure which control what is included implicitly. I want to avoid situations like #38251  I posted month ago. Also I think there was a thing where DTC_OVERLAY_FILE requires absolute path, but Kconfig overlays don't. I strongly believe name of the game should be consistency!

tejlmand committed:
cmake: kconfig: use absolute files for merge_config_files

The `merge_config_files` list contains files with both relative and
absolute paths.
Each entry is checked and files with a relative path are converted to
absolute path. This creates a `merge_config_files_with_absolute_paths`
identical to the `merge_config_files` but with absolute path on each
entry.

Afterwards it is mixed whether `merge_config_files` or
`merge_config_files_with_absolute_paths` are used creating unnecessary
risk when the list with relative path files are used.

This commit sets the content of `merge_config_files` with updated
absolute paths files, and also updates all places to use this list.

This ensures that after path conversion then only the list with absolute
paths is used.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
> Can I just chime in by saying whatever you decide on have it clearly documented somewhere.

Is this clear enough ?
https://github.com/zephyrproject-rtos/zephyr/blob/0c1077d8850b5d6291f97888334598cb9c8f1a61/doc/guides/build/kconfig/setting.rst#the-initial-configuration
> Per default all configuration files will be taken from the application's configuration directory except for files with an absolute path that are given with the CONF_FILE argument.
>
>The application's configuration directory will be defined as:
>
>  1.  If APPLICATION_CONFIG_DIR is specified by the user with `-DAPPLICATION_CONFIG_DIR=<path>` or in a CMake file before find_package(Zephyr) then this folder is used a the application's configuration directory.
>  1. If the folder :file:`boards/<BOARD>` exists, then this folder will be selected as the application's configuration directory.
>  1. The application's source directory.
>
> If a symbol is assigned both in :file:`<BOARD>_defconfig` and in the application configuration, the value set in the application configuration takes precedence.

SebastianBoe reviewed (changes requested):
Is this a breaking change for samples where \<app>/boards/\<board> exists ?

tejlmand commented:
>  a breaking change for samples

No.
For example you can still have:
```
<app>
  |--- prj.conf
  |--- boards
          |--- nrf52840dk_nrf52840.conf
```
in which case the `nrf52840dk_nrf52840.conf` will be merged with the `prj.conf` when building for the `nrf52840dk_nrf52850` board, just like today.

But if the `boards/nrf52840dk_nrf52840` is a folder, then it will be treated as the `APPLICATION_CONFIG_DIR` and config files will be picked up from there instead, for example.
```
<app2>
  |--- prj.conf                    # <--- Default prj.conf used, used for any other board.
  |--- boards
          |--- nrf52840dk_nrf52840
                  |--- prj.conf   # <--- prj.conf used for `nrf52840dk_nrf52850` board.
```

This also means that a downstream user can decide to only have:
```
<userapp>
  |--- boards
          |--- boardA
          |       |--- prj.conf
          |
          |--- boardB
                  |--- prj.conf
```
and that way indicate that this application can only be used with `boardA` or `boardB` because there is no common `prj.conf`.

Such a construct could also be usefull in Zephyr, for example in the IPC samples where there are constructs like this:
https://github.com/zephyrproject-rtos/zephyr/blob/ef0d955758e549f39d97b07089dedc4bdd05bc0e/samples/subsys/ipc/ipm_mcux/CMakeLists.txt#L9-L15
So that only supported boards will have a `boards/<supported-board>/prj.conf`.

Note, the IPC case still need a solution for the remote board to cleanup the CMakeLists.txt file, which is outside the scope of this PR. But this PR takes us one step closer in handling this kind of cases.

SebastianBoe commented:
No?

How can the prj.conf file change location without breaking the sample?

tejlmand commented:
> How can the prj.conf file change location without breaking the sample?

because it don't have to change location.
This feature is an opt-in.
All samples works the same as always per default with a `prj.conf` located directly under `<sample>/prj.conf`.

Only samples that want to use the new scheme can do so by creating a specific `prj.conf` for a given board under `boards/<board>/prj.conf` for that particular board.
That doesn't break how existing samples works.

If you are concerned for samples that might have a `boards/<board>/` folder already, then no Zephyr sample has that (already checked).

And we can't consider a downstream project that might have decided to create such a folder under boards a breaking change.
If we did so, then we could never introduce a new folder, cause somewhere downstream someone could already have been using such folder for custom purposes.

SebastianBoe commented:
> If you are concerned for samples that might have a boards/<board>/ folder already, then no Zephyr sample has that (already checked).
> 
> And we can't consider a downstream project that might have decided to create such a folder under boards a breaking change.
> 

That was my concern. And I do consider that a breaking change. If you don't then that is OK, the most important thing is that we have a common understanding of the actual semantics. Which we do now.

SebastianBoe commented on doc/guides/build/kconfig/setting.rst:
```diff
@@ -167,6 +167,22 @@ The application configuration can come from the sources below. By default,
 
+Per default all configuration files will be taken from the application's
+configuration directory except for files with an absolute path that are
+given with the ``CONF_FILE`` argument.
```
At line 156 we talk about the application directory.

But it should be application configuration directory, right?

Maybe even APPLICATION_CONFIGURATION_DIRECTORY to be clear that this is a precise term.

tejlmand replied:
We already have `APPLICATION_SOURCE_DIR` and `APPLICATION_BINARY_DIR` so I think sticking to `_DIR` fits nicely along that line. (which also includes CMAKE_CURRENT_SOURCE_DIR, CMAKE_CURRENT_LIST_DIR, etc.)

But with adjust to: `APPLICATION_CONFIGURATION_DIR` to be more precise.

And thanks for spotting the line 156, will fix.

SebastianBoe replied:
Sorry, used the wrong name.

SebastianBoe commented on doc/guides/build/kconfig/setting.rst:
```diff
@@ -167,6 +167,22 @@ The application configuration can come from the sources below. By default,
+2. If the folder :file:`boards/<BOARD>` exists, then this folder will be
+   selected as the application's configuration directory.
+
+3. The application's source directory.
```
Each of the three options for how an application configuration directory can be set should not mention that they are setting the application configuration directory.

At least, either all of them, or none of them should mention it. Now it is two of three.

SebastianBoe commented on doc/guides/build/kconfig/setting.rst:
```diff
@@ -167,6 +167,22 @@ The application configuration can come from the sources below. By default,
+configuration directory except for files with an absolute path that are
+given with the ``CONF_FILE`` argument.
+
+The application's configuration directory will be defined as:
```
APPLICATION_CONFIG_DIR will be set by one of the sources below with the highest priority listed first.

SebastianBoe commented on doc/guides/build/kconfig/setting.rst:
```diff
@@ -167,6 +167,22 @@ The application configuration can come from the sources below. By default,
+   configuration directory.
+
+2. If the folder :file:`boards/<BOARD>` exists, then this folder will be
+   selected as the application's configuration directory.
```
2. The directory :file:`boards/<BOARD>`, if it exists.

tejlmand commented:
>  the most important thing is that we have a common understanding of the actual semantics. Which we do now.

Agree. All comments should be addressed now.

galak reviewed (commented):
dev-review (Oct 7, 2021):
* pull out the `boards/<BOARD>` into a separate PR, its a change orthogonal to the APPLICATION_CONFIG_DIR

mbolivar-nordic reviewed (approved):
Just an ACK assuming this works the way we discussed in dev-review
